sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/package.scala

```diff
@@ -81,4 +85,221 @@ package object state {
+   *
+   * This implementation only works when input row attributes contain all the key attributes.
+   *
+   * @param keyExpressions The attributes of keys.
+   * @param inputRowAttributes The attributes of input row.
+   */
+  class StreamingAggregationStateManagerImplV1(
+      keyExpressions: Seq[Attribute],
+      inputRowAttributes: Seq[Attribute])
+    extends StreamingAggregationStateManagerBaseImpl(keyExpressions, inputRowAttributes) {
+
+    override def getStateValueSchema: StructType = inputRowAttributes.toStructType
+
+    override def get(store: StateStore, key: UnsafeRow): UnsafeRow = {
+      store.get(key)
+    }
+
+    override def put(store: StateStore, row: UnsafeRow): Unit = {
+      store.put(getKey(row), row)
+    }
+
+    override def iterator(store: StateStore): Iterator[UnsafeRowPair] = {
+      store.iterator()
+    }
+
+    override def values(store: StateStore): Iterator[UnsafeRow] = {
+      store.iterator().map(_.value)
+    }
+  }
+
+  /**
+   * The implementation of StreamingAggregationStateManager for state version 2.
+   * In state version 2, the schema of key and value in state are follow:
+   *
+   * - key: Same as key expressions.
+   * - value: The diff between input row attributes and key expressions.
+   *
+   * The schema of value is changed to optimize the memory/space usage in state, via removing
+   * duplicated columns in key-value pair. Hence key columns are excluded from the schema of value.
+   *
+   * This implementation only works when input row attributes contain all the key attributes.
+   *
+   * @param keyExpressions The attributes of keys.
+   * @param inputRowAttributes The attributes of input row.
+   */
+  class StreamingAggregationStateManagerImplV2(
+      keyExpressions: Seq[Attribute],
+      inputRowAttributes: Seq[Attribute])
+    extends StreamingAggregationStateManagerBaseImpl(keyExpressions, inputRowAttributes) {
+
+    private val valueExpressions: Seq[Attribute] = inputRowAttributes.diff(keyExpressions)
+    private val keyValueJoinedExpressions: Seq[Attribute] = keyExpressions ++ valueExpressions
+    private val needToProjectToRestoreValue: Boolean =
+      keyValueJoinedExpressions != inputRowAttributes
+
+    @transient private lazy val valueProjector =
+      GenerateUnsafeProjection.generate(valueExpressions, inputRowAttributes)
+
+    @transient private lazy val joiner =
+      GenerateUnsafeRowJoiner.create(StructType.fromAttributes(keyExpressions),
+        StructType.fromAttributes(valueExpressions))
+    @transient private lazy val restoreValueProjector = GenerateUnsafeProjection.generate(
+      keyValueJoinedExpressions, inputRowAttributes)
+
+    override def getStateValueSchema: StructType = valueExpressions.toStructType
+
+    override def get(store: StateStore, key: UnsafeRow): UnsafeRow = {
+      val savedState = store.get(key)
+      if (savedState == null) {
+        return savedState
+      }
+
+      val joinedRow = joiner.join(key, savedState)
```

Comment:
cant you dedup the code with `restoreOriginRow` method?

Reply:
Missed spot. Will leverage `restoreOriginRow`.